### PR TITLE
#165273571 Social authentication

### PR DESCRIPTION
**Pull request by JamesMudidi**
#### What does this PR do?
Sets up Social authentication

#### Description of Tasks to be completed?
Added authentication for the following social platforms
- Facebook
- Google

#### How should this be manually tested?
After cloning the repository,

- Navigate into it and install the dependencies using `pip3 install -r requirements.txt`
- Setup a database with PostgreSQL
- Set up your environment variables following the sample.env file provided and then run `source .env`
- Make migrations using `python manage.py makemigrations`
- Appl the migrations using `python manage.py migrate`
- Run the application using `python3 manage.py runserver`
- Create a developer account on any of the social media platforms supported
- Start postman and access the desired endpoint using `http://127.0.0.1:8000/api/auth/<social platform>/`. where <social platform> can be facebook
- Generate and copy the token from the already created developer account of the desired social platform eg. facebook
- use the token as shown in the screenshot below

#### Any background context you want to provide?
This feature works with third-party Apis so you need to have some knowledge on the use of those Apis before proceeding to test the feature. The links below will help you get started with creating an account to become a developer for Facebook and Google where you can generate tokens and test the API.
- [Facebook](https://developers.facebook.com/)
- [Google](https://developers.google.com/oauthplayground/)

#### Screenshots

Facebook Screenshoot
Screenshot 2019-05-07 at 12 11 40 PM

Google ScreenShoot
Screenshot 2019-05-07 at 12 13 15 PM

#### What are the relevant pivotal tracker stories?
[165273571](https://www.pivotaltracker.com/story/show/165273571)

### CODE REVIEW

**Review by bisonlou — DISMISSED** [review]
Good work @JamesMudidi. I have tested google authentication.

File: authors/apps/authentication/social.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def register_social_user(email, username):
+    user = User.objects.filter(email=email)
+
+    if user.exists():
+        current_user = User.objects.get(email=email)
+        current_user.is_active = True
+        current_user.save()
+        registered_social_user = authenticate(email=email, password="Andelatesting2")
```
Comment (bisonlou) [inline authors/apps/authentication/social.py]: I notice the password is plain for all to see. Does this mean that as long as I know the email of a user (which is public to everyone) I have access to the user's account once they register using social auth?

File: authors/apps/authentication/social.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        '''At this point of the GoogleHandler class we are
+        Getting and Validating the user information from
+        the authorization token'''
+
+        try:
+            idinfo = id_token.verify_oauth2_token(
+                auth_token, requests.Request())
+            return idinfo
+        except Exception:
+            return "he token provided is invalid or has expired"
```
Comment (bisonlou) [inline authors/apps/authentication/social.py]: I believe there is a typo here. Correct me if am wrong.

**Review by bisonlou — APPROVED** [review]

File: authors/apps/authentication/social.py (hunk 1 above)
Comment (bisonlou) [inline authors/apps/authentication/social.py]: I notice the password is plain for all to see. Does this mean that as long as I know the email of a user (which is public to everyone), I have access to the user's account once they register using social auth?

File: authors/apps/authentication/social.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    handle the user information got from google'''
+
+    @staticmethod
+    def validate(auth_token):
+        '''At this point of the GoogleHandler class we are
+        Getting and Validating the user information from
+        the authorization token'''
+
+        try:
+            idinfo = id_token.verify_oauth2_token(
```
Comment (bisonlou) [inline authors/apps/authentication/social.py]: this is not a very meaningful variable name. Please consider changing it to id_info.

File: authors/apps/authentication/social.py (hunk 2 above)
Comment (bisonlou) [inline authors/apps/authentication/social.py]: I believe this is a typo.

**Review by bisonlou — CHANGES_REQUESTED** [review]
Great work @JamesMudidi. Though I am really concerned about the security loophole created by having a known password set for all users using social auth.

File: authors/apps/authentication/social.py (hunk 1 above)
Comment (kalsmic) [reply authors/apps/authentication/social.py]: @JamesMudidi, The default password for social auth authentication should be stored in an environmental variable.

**Review by bisonlou — APPROVED** [review]

**Review by ZackAtama — APPROVED** [review]
Great work @JamesMudidi

**Review by llwasampijja — APPROVED** [review]

**Review by nadralia — APPROVED** [review]
@JamesMudidi Nicework

**Pull request merged**